cmdstanpy/model.py

```diff
@@ -1302,18 +1337,32 @@ def _run_cmdstan(
         Args 'show_progress' and 'show_console' allow use of progress bar,
         streaming output to console, respectively.
         """
-
-        cmd = runset.cmds[idx]
         get_logger().debug(
             'threads: %s', str(os.environ.get('STAN_NUM_THREADS'))
         )
-        if show_progress and progbar.allow_show_progress():
-            progress_hook: Optional[
-                Callable[[str], None]
-            ] = self._wrap_sampler_progress_hook(idx + 1, iter_total)
-        else:
-            get_logger().info('start chain %d', idx + 1)
-            progress_hook = None
+        get_logger().debug('runset: %s', runset.__repr__())
+
+        logger_prefix = 'CmdStan'
+        console_prefix = ''
+        if runset.one_process_per_chain:
+            logger_prefix = 'Chain [{}]'.format(idx + 1)
+            console_prefix = 'Chain [{}] '.format(idx + 1)
+        if show_console:
+            get_logger().info('%s start processing', logger_prefix)
+
+        progress_hook: Optional[Callable[[str], None]] = None
+        if show_progress and runset.one_process_per_chain:
+            progress_hook = self._wrap_sampler_per_chain_progress_hook(
+                chain_id=idx + 1, total=iter_total
+            )
+        elif show_progress:
+            progress_hook = self._wrap_sampler_multi_chain_progress_hook(
+                num_chains=runset.chains,
+                offset=runset.chain_ids[0],
+                total=iter_total,
+            )
```

Comment:
Is there a reason to prefer this over using one callback that handles both cases?

Reply:
much easier to implement - you've fundamentally got two different data structures to manage - a single progress bar or an array of progress bars.  if you have a single method with two different data structures, it's really easy to mess up on the update statements.

I can take a closer look at the errors/exceptions thrown by the tqdm library and see if there's a more elegant solution.  In my option, it's whack-a-mole on understandability and maintainability here.

Reply:
My partial solution was just to always use the “list of progress bars”, even in the one-chain-one-process case. You have to make the hook a parameter of the _run_cmdstan method, but since it’s only used by one of the methods that almost makes more sense IMO.

Reply:
refactored into a single method.  the conditional logic is one thing, the bookkeeping is another - the combination is maddening.